clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -128,6 +128,161 @@ static Attr *handleIntelFPGALoopAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
+    SafelenValue = ArgVal.getZExtValue();
+    return IVDepExprResult::SafeLen;
+  }
+
+  if (isa<DeclRefExpr>(E)) {
+    if (!cast<DeclRefExpr>(E)->getType()->isArrayType()) {
+      S.Diag(E->getExprLoc(), diag::err_ivdep_declrefexpr_arg);
+      return IVDepExprResult::Invalid;
+    }
+    return IVDepExprResult::Array;
+  }
+
+  S.Diag(E->getExprLoc(), diag::err_ivdep_unknown_arg);
+  return IVDepExprResult::Invalid;
+}
+
+// Note: At the time of this call, we don't know the order of the expressions,
+// so we name them vaguely until we can figure it out.
+SYCLIntelFPGAIVDepAttr *
+Sema::BuildSYCLIntelFPGAIVDepAttr(const AttributeCommonInfo &CI, Expr *Expr1,
+                                  Expr *Expr2) {
+  unsigned SafelenValue = 0;
+  IVDepExprResult E1 = HandleFPGAIVDepAttrExpr(*this, Expr1, SafelenValue);
+  IVDepExprResult E2 = HandleFPGAIVDepAttrExpr(*this, Expr2, SafelenValue);
+
+  if (E1 == IVDepExprResult::Invalid || E2 == IVDepExprResult::Invalid)
+    return nullptr;
+
+  if (E1 == E2 && E1 != IVDepExprResult::Dependent &&
+      E1 != IVDepExprResult::Null) {
+    Diag(Expr2->getExprLoc(), diag::err_ivdep_duplicate_arg);
+    return nullptr;
+  }
+
+  // Try to put Safelen in the 1st one so codegen can count on the ordering.
+  Expr *SafeLenExpr;
+  Expr *ArrayExpr;
+  if (E1 == IVDepExprResult::SafeLen) {
+    SafeLenExpr = Expr1;
+    ArrayExpr = Expr2;
+  } else {
+    SafeLenExpr = Expr2;
+    ArrayExpr = Expr1;
+  }
+
+  return new (Context)
+      SYCLIntelFPGAIVDepAttr(Context, CI, SafeLenExpr, ArrayExpr, SafelenValue);
+}
+
+static void
+CheckRedundantSYCLIntelFPGAIVDepAttrs(Sema &S, ArrayRef<const Attr *> Attrs) {
+  // Skip SEMA if we're in a template, this will be diagnosed later.
+  if (S.getCurLexicalContext()->isDependentContext())
+    return;
+
+  SmallVector<const SYCLIntelFPGAIVDepAttr *, 8> FilteredAttrs;
+  // Filter down to just non-dependent ivdeps.
+  llvm::transform(Attrs, std::back_inserter(FilteredAttrs), [](const Attr *A) {
+    if (const auto *IVDep = dyn_cast_or_null<const SYCLIntelFPGAIVDepAttr>(A))
+      return IVDep->isDependent() ? nullptr : IVDep;
+    return static_cast<const SYCLIntelFPGAIVDepAttr *>(nullptr);
+  });
+  FilteredAttrs.erase(
+      std::remove(FilteredAttrs.begin(), FilteredAttrs.end(),
+                  static_cast<const SYCLIntelFPGAIVDepAttr *>(nullptr)),
+      FilteredAttrs.end());
+  if (FilteredAttrs.empty())
+    return;
+
+  SmallVector<const SYCLIntelFPGAIVDepAttr *, 8> SortedAttrs(FilteredAttrs);
+  llvm::stable_sort(SortedAttrs, SYCLIntelFPGAIVDepAttr::SafelenCompare);
+
+  // Find the maximum without an array expression, which ends up in the 2nd
+  // expr.
+  const auto *GlobalMaxItr =
+      llvm::find_if(SortedAttrs, [](const SYCLIntelFPGAIVDepAttr *A) {
+        return !A->getArrayExpr();
+      });
+  const SYCLIntelFPGAIVDepAttr *GlobalMax =
+      GlobalMaxItr == SortedAttrs.end() ? nullptr : *GlobalMaxItr;
+
+  for (const auto *A : FilteredAttrs) {
```

Comment:
I use the sorted attrs to make sure I always can easily find the correct value.  I need to use FilteredAttrs so that I maintain declaration order, otherwise the diagnostics end up being in the wrong order.

The loop itself depends on getting the attributes in declaration order, but needs to have a sorted list such that it can refer to it to find the 'greatest' of each array type.